src/lib.rs

```diff
@@ -0,0 +1,147 @@
+#[cfg(feature = "serde")]
+#[macro_use]
+extern crate serde;
+extern crate num_traits;
+
+use num_traits::{AsPrimitive, FromPrimitive, PrimInt, ToPrimitive, Unsigned};
+use packed_vec::PackedVec;
+use vob::Vob;
+
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct SparseVec<T> {
+    displacement: Vec<usize>,
+    row_length: usize,
+    empty_val: T,
+    empties: Vob,
+    data: PackedVec<T>,
+}
+
+impl<T: Clone + Copy + PartialEq> SparseVec<T>
+where
+    T: AsPrimitive<usize> + FromPrimitive + Ord + PrimInt + ToPrimitive + Unsigned,
+    usize: AsPrimitive<T>,
+{
+    pub fn from(v: &Vec<T>, empty_val: T, row_length: usize) -> SparseVec<T> {
+        if v.len() == 0 {
+            return SparseVec {
+                displacement: Vec::new(),
+                row_length: 0,
+                empty_val,
+                empties: Vob::new(),
+                data: PackedVec::new(v.to_vec()),
+            };
+        }
+
+        let (c, d) = compress(v, empty_val, row_length);
+        let e = calc_empties(v, empty_val);
+        let pv = PackedVec::new(c);
+        SparseVec {
+            displacement: d,
+            row_length,
+            empty_val,
+            empties: e,
+            data: pv,
+        }
+    }
+
+    pub fn get(&self, r: usize, c: usize) -> Option<T> {
+        let k = r * self.row_length + c;
+        if k >= self.empties.len() {
+            return None;
+        }
+        if self.empties.get(k).unwrap() {
+            Some(self.empty_val)
+        } else {
+            let d = self.displacement[r];
+            self.data.get(d + c)
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.empties.len()
+    }
+}
+
+fn calc_empties<T: PartialEq>(vec: &Vec<T>, empty_val: T) -> Vob {
+    let mut vob = Vob::from_elem(vec.len(), false);
+    for i in 0..vec.len() {
+        if vec[i] == empty_val {
+            vob.set(i, true);
+        }
+    }
+    vob
+}
+
+fn compress<T: Clone + Copy + PartialEq>(
+    vec: &Vec<T>,
+    empty_val: T,
+    row_length: usize,
+) -> (Vec<T>, Vec<usize>) {
+    let mut r = Vec::new(); // Result vector
+    r.resize(row_length, empty_val);
+
+    let mut d = 0; // displacement value
+    let mut dv = Vec::new(); // displacement vector
+
+    for s in vec.chunks(row_length) {
+        loop {
+            if fits(&s, &r, d, empty_val) {
+                apply(s, &mut r, d, empty_val);
+                dv.push(d);
+                break;
+            } else {
+                d += 1;
```

Comment:
And probably this to be `d += row_length` too?

Reply:
OK, so I think I understand this. I'm not sure I understand the `resize` bit though.

Reply:
I just realised this has a bug. The idea was that whenever a vector can't be merged and needs to be shifted, we need to increase the size of compressed vector, to fit the shifted vector in.

However, I just realised that since I'm not resetting `d`, I'm not trying later vectors at earlier positions, and so might be shifting some later vectors even if they would fit at position 0.
So I only need to increase the result vector size, if the `d + row_length > r.len()`.